crawl-ref/source/player.cc

```diff
@@ -3478,6 +3512,36 @@ void dec_mp(int mp_loss, bool silent)
         flush_mp();
 }
 
+void pay_mp(int cost)
+{
+    if (you.has_mutation(MUT_HP_CASTING))
```

Comment:
Great question! There's a very important distinction being made here. `pay_mp()` covers MP *costs*, for spells, abilities, evocables, etc. But there are also external effects that decrease your MP - ghost moth drain gaze, zot trap drain MP effect, etc. We *don't* want those to decrease your HP, so we want a different codepath there - `dec_mp`. Does that make sense?

Reply:
It still seems like a fair amount of working around the pay vs dec distinction in this patch that will lead to confusion. I wonder if it would ultimately be simpler to think of Dj/innate casters as immune to specific draining effects and special-case those? It really doesn't look like a lot of places (beam.cc, god-wrath.cc, traps.cc), if anything less special casing than the current patch; perhaps wrap those into a separate function as opposed to making the pay vs decrement distinction?

Reply:
Per our discussion in IRC, I've gone with a `drain_mp` function and a `pay_mp` function.